pkg/controller/jobs/statefulset/statefulset_reconciler.go

```diff
@@ -134,6 +138,43 @@ func shouldFinalize(sts *appsv1.StatefulSet, pod *corev1.Pod) bool {
 	return shouldUngate(sts, pod) || utilpod.IsTerminated(pod)
 }
 
+func (r *Reconciler) reconcileWorkload(ctx context.Context, sts *appsv1.StatefulSet) error {
+	if sts == nil {
+		return nil
+	}
+
+	wl := &kueue.Workload{}
+	err := r.client.Get(ctx, client.ObjectKey{Namespace: sts.Namespace, Name: GetWorkloadName(sts.Name)}, wl)
+	if err != nil {
+		return client.IgnoreNotFound(err)
+	}
+
+	hasOwnerReference, err := controllerutil.HasOwnerReference(wl.OwnerReferences, sts, r.client.Scheme())
+	if err != nil {
+		return err
+	}
+
+	var (
+		shouldUpdate = false
+		replicas     = ptr.Deref(sts.Spec.Replicas, 1)
+	)
+
+	switch {
+	case hasOwnerReference && replicas == 0:
+		shouldUpdate = true
+		err = controllerutil.RemoveOwnerReference(sts, wl, r.client.Scheme())
```

Comment:
Can you describe why we remove the owner reference from the workload on scaling down the StatefulSet?

IIUC this will result in deleting the workload, right? It seems safer to just update the count in the workload object, similarly as we would do for Jobs. For example, when we suspend a Job we don't delete the workload IIRC.

Reply:
> For example, when we suspend a Job we don't delete the workload IIRC.

It is a very interesting behavior in Job. I tried to suspend the job manually, but it automatically unsuspends itself. The only way to suspend the job is to deactivate the workload.

Reply:
> Can you describe why we remove the owner reference from the workload on scaling down the StatefulSet?

The problem is that after activation, the replicas can change (replicas=3 → replicas=0 → replicas=5). Our workload is admitted, and we can't change the PodSet count since this field is immutable. I think it's much easier to just delete it.

Reply:
> IIUC this will result in deleting the workload, right?

Yeah, you right. It should delete workload.

Reply:
Yeah, for Job when the count is updated, then we update the count in workload. 

So, my natural preference is to also update the count in the workload for STS, not sure it is much more complex. I'm not sure myself - considering the options.

 > The problem is that after activation, the replicas can change (replicas=3 → replicas=0 → replicas=5). Our workload is admitted, and we can't change the PodSet count since this field is immutable. I think it's much easier to just delete it.

I see, but to me it suggests we should also evict the workload for STS in that case, to be consistent with Job.
Can you try to prototype, maybe in a separate PR so that we can compare. I don't think we need to rush with decisions. I would prefer to do it well.

Reply:
Thank you for the investigation and answers. The PR looks good assuming we take the approach of deleting the workload, but I would prefer to explore keeping and updating the workload, as we do for Jobs.